### PR TITLE
[WIP] ATG-Console (Automatic Test Generation Console)

### PR DESCRIPTION
**Pull request by Aditya-eddy**
Console provided to show demo for keploy to user.

### Design a User-Friendly Interface:
Create a visually appealing and intuitive layout that allows users to easily navigate and understand the test generation features. Collaborate with mentors to finalize the design.

### Implement Test Record and Replay:
Integrate functionality to record user actions and replay them as tests, demonstrating the ease of creating tests with Keploy.

### Showcase Automatic Test Generation:
Highlight how Keploy can automatically generate tests based on user interactions, showcasing its efficiency and effectiveness.

### Interactive Demo:
Develop an interactive demo that allows users to experience firsthand how Keploy simplifies the test generation process.

### Documentation and Tutorials:
Include comprehensive documentation and tutorials within the UI to guide users through the demo and explain the underlying concepts and benefits of using Keploy.

### Responsive Design:
Ensure the UI is responsive and accessible across various devices and screen sizes, providing a seamless user experience for everyone.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (nehagup) [inline .vscode/launch.json:1]: you can add this to git-ignore

Comment (Aditya-eddy) [reply .vscode/launch.json:1]: okay!!

File: tailwind.config.js [hunk 2]
```diff
@@ -1,220 +1,226 @@
...
-            spacing: {
-                '9/16': '56.25%',
-                '3/4': '75%',
-                '1/1': '100%',
-            },
-            fontFamily: {
-                inter: ['var(--font-inter)', 'sans-serif'],
-                'architects-daughter': ['var(--font-architects-daughter)', 'sans-serif'],
-                pencil: ['Pencil', 'sans-serif'],
+  content: [
```
Comment (nehagup) [inline tailwind.config.js]: can you please remove the linting changes, will make it easier for review.

File: app/(default)/playground/page.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import React from "react";
+import Editor from "@/components/atg/editor";
```
Comment (charankamarapu) [inline app/(default)/playground/page.tsx:2]: what is atg..? Please change it to a sutiable name.

File: app/(default)/playground/page.tsx [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+const page = () => {
+  return (
+    <div className="mt-24">
+      <Editor goFullScreen={true} />
```
Comment (charankamarapu) [inline app/(default)/playground/page.tsx:6]: Here too please change the name from editor to something else.

File: app/(default)/test-case-generator/page.tsx [hunk 5]
```diff
@@ -10,23 +10,28 @@ import { Testimonial } from "@/components/testimonial";
...
+  keywords: "test case generation",
 };
 
 const heroData = {
   titleTop: "Improve Tests with ",
   titleBottom: "E2E Test Generator",
   subtitleTop: "Avoid confusion and Duplication in Test Execution!",
   subtitleBottom:
-    "Remove the duplicate and redundant test cases in minutes. 🚀",
+    "Remove the duplicate and redundant test cases in minutes. Try our new Automatic test Generator 🚀",
```
Comment (charankamarapu) [inline app/(default)/test-case-generator/page.tsx:33]: why is this sentence in deduce subtitle...?

File: app/(default)/test-case-generator/page.tsx [hunk 6]
```diff
@@ -10,23 +10,28 @@ import { Testimonial } from "@/components/testimonial";
...
-import { featuresData,faqQuestions } from "@/components/utils/test-case-generator";
+import {
+  featuresData,
+  faqQuestions,
+} from "@/components/utils/test-case-generator";
 import BannerBunny from "@/public/images/banner-bunny.png";
 import FAQ from "@/components/pillar-page/faq";
 import RootLayout, { Metadata } from "@/app/layout";
-
+import Editor from "@/components/atg/editor";
```
Comment (charankamarapu) [inline app/(default)/test-case-generator/page.tsx]: is this used in this page because I can see that is commented?

File: app/(default)/test-case-generator/page.tsx [hunk 7]
```diff
@@ -60,8 +65,11 @@ export default function Home() {
         titleBottom={heroData.titleBottom}
         subtitleTop={heroData.subtitleTop}
         subtitleBottom={heroData.subtitleBottom}
-        animationSrc={heroData.animationSrc}
+        editor={true}
+        extraClasses="text-center" // Pass this to ensure text is centered
       />
+
+      {/* <Editor/> */}
```
Comment (Hermione2408) [inline app/(default)/test-case-generator/page.tsx]: remove

File: app/api/hello/atg.ts
Comment (Hermione2408) [inline app/api/hello/atg.ts:1]: why api/hello ?

File: app/api/hello/atg.ts [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+  language: string;
+  schema: string;
+  code: string;
+}) => {
+  try {
+    const headers = {
+      "Content-Type": "application/json",
+    };
+
+    const requestBody = {
```
Comment (Hermione2408) [inline app/api/hello/atg.ts]: why aren't you using useMutation hook in apollo graphql instead of fetch ?

File: app/api/hello/atg.ts [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+};
+
+type RunCommandSubscriptionParams = {
+  codeSubmissionId: string;
+  command: string;
+  testSetName?: string; // Optional parameter
+};
+
+const RUN_COMMAND_SUBSCRIPTION = gql`
+  subscription RunCommand(
```
Comment (Hermione2408) [inline app/api/hello/atg.ts]: will be good to create a separate file for gql queries/subscription/mutation definitions

File: app/api/hello/atg.ts [hunk 10]
```diff
@@ -0,0 +1,337 @@
+import { gql, useSubscription } from "@apollo/client";
```
Comment (charankamarapu) [inline app/api/hello/atg.ts]: only keep what are required in the component Things like curl query etc can be removed so that we can review easily. This file has to be in services.

File: app/api/hello/atg.ts [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+  testSetName: string
+): Promise<CommandResponse> {
+  const query = `
+    subscription FetchMock($code_submission_id: String!, $command: String!, $test_set_name: String!) {
+      runCommand(code_submission_id: $code_submission_id, command: $command, test_set_name: $test_set_name)
+    }
+  `;
+  const variables = {
+    code_submission_id: codeSubmissionId,
+    command: "FETCH_MOCK",
```
Comment (Hermione2408) [inline app/api/hello/atg.ts]: we can have enum for all commands and use that

File: app/api/hello/atg.ts [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+    test_set_report_name: testSetReportName // Updated to match the query
+  };
+  return await postRequest(
+    "https://landing-page.staging.keploy.io/query",
+    query,
+    variables
+  );
+}
+
+export async function RemovingDuplicate(
```
Comment (Hermione2408) [inline app/api/hello/atg.ts]: inconsistent function naming

File: components/atg/Terminal/TerminalComponents/DedupulicationTerminal.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
+import React, {
+  useMemo,
+  useEffect,
```
Comment (charankamarapu) [inline components/atg/Terminal/TerminalComponents/DedupulicationTerminal.tsx]: what are these separate terminals for deduce, record etc

File: components/atg/Utils/Schema.ts [hunk 14]
```diff
@@ -0,0 +1,8 @@
+export const GolangSchema =
```
Comment (charankamarapu) [inline components/atg/Utils/Schema.ts]: why are these here, should these be fetched..?

File: components/atg/data/Python.ts [hunk 15]
```diff
@@ -0,0 +1,48 @@
+import { Type, Directory } from "../Editor/utils/file-manager";
```
Comment (charankamarapu) [inline components/atg/data/Python.ts]: shouldn't these data be brought from demo-project repo...?

**Review by charankamarapu — CHANGES_REQUESTED** [review]
Please address the comments

File: components/TestingConsole.tsx
Comment (Hermione2408) [inline components/TestingConsole.tsx:1]: what's the point of this file if its simply returning other component without any changes

File: components/atg/Utils/Schema.ts
Comment (Hermione2408) [inline components/atg/Utils/Schema.ts:1]: utils -> in small case, similary for schema.ts

File: components/atg/Editor/components/file-tree.tsx [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+const SubTree = (props: SubTreeProps) => {
+  const [emptyDirectory, setEmptyDirectory] = useState<boolean>(false);
+  return (
+    <div>
+      {emptyDirectory ? (
+        <div className="h-2 flex items-center "></div>
+      ) : (
+        <>
+          {props.directory.dirs.sort(sortDir).map((dir) => (
```
Comment (Hermione2408) [inline components/atg/Editor/components/file-tree.tsx]: You could create the processed array of components outside the return statement and then just render it inside the JSX. This separates the logic of "what to display" from "how to display it."

File: components/atg/Editor/components/file-tree.tsx [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+            <DirDiv
+              key={dir.id}
+              directory={dir}
+              selectedFile={props.selectedFile}
+              onSelect={props.onSelect}
+              themeDir={props.themeSub}
+            />
+          ))}
+          {props.directory.files.sort(sortFile).map((file) => (
```
Comment (Hermione2408) [inline components/atg/Editor/components/file-tree.tsx]: create array outside return then map here

File: components/atg/Editor/components/file-tree.tsx
Comment (Hermione2408) [inline components/atg/Editor/components/file-tree.tsx:1]: why are there multiple components in same file, please have separate file for separate components

File: components/atg/Editor/components/file-tree.tsx [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+  themeFileDiv,
+}: {
+  file: File | Directory;
+  icon?: string;
+  selectedFile: File | undefined;
+  onClick: () => void;
+  openParameter?: boolean;
+  themeFileDiv: boolean;
+}) => {
+  const isSelected = selectedFile && selectedFile.id === file.id;
```
Comment (Hermione2408) [inline components/atg/Editor/components/file-tree.tsx]: wrap it with Boolean, Wrapping the isSelected variable with Boolean ensures it is always strictly true or false, preventing potential issues with non-boolean values like undefined or null

File: components/atg/Editor/components/file-tree.tsx [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+  themeDir: boolean;
+}) => {
+  const [open, setOpen] = useState(() =>
+    isChildSelected(directory, selectedFile)
+  );
+  const [dirState, setDirState] = useState(directory);
+  useState(() => {
+    if (
+      directory.id === "src" ||
+      directory.id === "src2" ||
```
Comment (Hermione2408) [inline components/atg/Editor/components/file-tree.tsx]: add comment for why is this done

File: components/atg/Editor/components/file-tree.tsx [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+      directory.id === "src3"
+    ) {
+      setOpen(true);
+    }
+  });
+
+  const depth = directory.depth;
+
+  const handleToggle = async () => {
```
Comment (Hermione2408) [inline components/atg/Editor/components/file-tree.tsx]: why async? handleToggle simply toggles a boolean state (setOpen(!open);) without performing any asynchronous operations, such as API calls, timeouts, or other await expressions.

File: components/atg/Terminal/index.tsx [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+import { TerminalProps, TerminalCommands } from "./types"; // Assuming you have a types file where `TerminalCommands` is defined
+import CloseIcon from "@mui/icons-material/Close"; // Use CloseIcon to mimic the red button on Mac
+
+export interface AdditionalProps {
+  inputRef: React.RefObject<HTMLInputElement>;
+  terminalTheme: boolean;
```
Comment (Hermione2408) [inline components/atg/Terminal/index.tsx:13]: how is terminalTheme boolean ? shouldnt it be a string or is it named incorrectly

File: components/atg/Terminal/index.tsx [hunk 22]
```diff
@@ -0,0 +1,289 @@
...
+    // Function to handle command not found
+    const handleCommandNotFound = () => {
+      const commandNotFoundFunc = commands["__notFound__"];
+      if (commandNotFoundFunc) {
+        commandNotFoundFunc();
+      }
+    };
+
+    /**
+     * Update the input width based on the span width
```
Comment (Hermione2408) [inline components/atg/Terminal/index.tsx]: remove

File: components/atg/Terminal/index.tsx [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+      terminalTheme,
+      SetTerminalHeight,
+      Loading,
+    } = props;
+
+    const [input, setInputValue] = useState<string>("");
+    const [blinkingTrue, setBlinkingTrue] = useState<boolean>(true);
+    const spanRef = useRef<HTMLSpanElement>(null);
+    const typingTimeoutRef = useRef<NodeJS.Timeout | null>(null);
+    const [height, setHeight] = useState<string>("orange");
```
Comment (Hermione2408) [inline components/atg/Terminal/index.tsx:36]: incorrect naming "orange" for height ?

File: components/atg/Utils/Schema.ts
Comment (Hermione2408) [inline components/atg/Utils/Schema.ts:1]: is this file needed, as there were no reference for variables defined

**Review by Hermione2408 — CHANGES_REQUESTED** [review]

File: app/(default)/test-case-generator/page.tsx (hunk 6 above)
Comment (Aditya-eddy) [reply app/(default)/test-case-generator/page.tsx]: Removed it!

File: components/atg/Terminal/TerminalComponents/DedupulicationTerminal.tsx (hunk 13 above)
Comment (Aditya-eddy) [reply components/atg/Terminal/TerminalComponents/DedupulicationTerminal.tsx]: We had to create seperate terminals because specific text-based loading was happening in side-bar which would a little complex for a  single terminal. As we had to run TEST_GENERATE query , DEDUP query etc. So having single terminal would increase the complexity for it. But if we had to create a single-file for the terminal i can do that!

File: components/atg/Utils/Schema.ts (hunk 14 above)
Comment (Aditya-eddy) [reply components/atg/Utils/Schema.ts]: These schemas are present are defined in the rootDir. So for the component in which it is being used instead of passing the whole of rootDir (which contains all of files). It would be heavy prop. So i made a seprate schema file which we can use.

File: components/atg/data/Python.ts (hunk 15 above)
Comment (Aditya-eddy) [reply components/atg/data/Python.ts]: No , I made a file-structure for  every language as .ts file in /atg/data folder. Should we fetch them directly from the github repo?

File: app/(default)/test-case-generator/page.tsx (hunk 7 above)
Comment (Aditya-eddy) [reply app/(default)/test-case-generator/page.tsx]: removed it!

File: app/api/hello/atg.ts (hunk 8 above)
Comment (Aditya-eddy) [reply app/api/hello/atg.ts]: Because from the apollo client what i wanted was the useSubscription hook only. Rest Api's can be just simply fetched using fetch as it is just a post request. Should i convert them to using useMutation hook?

File: components/atg/Editor/components/file-tree.tsx (hunk 20 above)
Comment (Aditya-eddy) [reply components/atg/Editor/components/file-tree.tsx]: Yup removing it right now!!

File: components/TestingConsole.tsx
Comment (Aditya-eddy) [reply components/TestingConsole.tsx:1]: removed it!!

File: components/atg/Utils/Schema.ts
Comment (Aditya-eddy) [reply components/atg/Utils/Schema.ts:1]: made file name in small case.

File: components/atg/Editor/components/file-tree.tsx (hunk 19 above)
Comment (Aditya-eddy) [reply components/atg/Editor/components/file-tree.tsx]: added the comment for it. Basically it was for when the file tree renders it's top directory which is src. It should render open meaning all of the files will be displayed rather than user going all the way to click a button and then the file-tree would open.

File: components/atg/Terminal/index.tsx (hunk 22 above)
Comment (Aditya-eddy) [reply components/atg/Terminal/index.tsx]: removed it!

File: components/atg/Utils/Schema.ts
Comment (Aditya-eddy) [reply components/atg/Utils/Schema.ts:1]: Yup this file is needed correcting it!

File: components/atg/Terminal/index.tsx (hunk 21 above)
Comment (Aditya-eddy) [reply components/atg/Terminal/index.tsx:13]: terminalTheme is boolean. If it's true then it is light theme if it false it is dark mode. Let me correct the naming.

File: app/api/hello/atg.ts (hunk 8 above)
Comment (Hermione2408) [reply app/api/hello/atg.ts]: Apollo's useMutation hook is designed specifically for executing GraphQL mutations. By using it, you keep all your GraphQL-related logic consistent, whether it's queries, mutations, or subscriptions.You already setup graphql client so why are you not using it for mutations?

File: components/atg/data/Python.ts (hunk 15 above)
Comment (charankamarapu) [reply components/atg/data/Python.ts]: yes

File: components/atg/Utils/Schema.ts (hunk 14 above)
Comment (charankamarapu) [reply components/atg/Utils/Schema.ts]: fetch from the repo.

File: components/atg/Terminal/TerminalComponents/DedupulicationTerminal.tsx (hunk 13 above)
Comment (charankamarapu) [reply components/atg/Terminal/TerminalComponents/DedupulicationTerminal.tsx]: will discuss this offline.

File: app/api/createFileStructure/route.ts [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+const pathsMap: { [key: string]: string } = {
+  Golang: 'components/atg/demo-projects/projects/goapp',
+  Javascript: 'components/atg/demo-projects/projects/javaScript',
+  Python: 'components/atg/demo-projects/projects/python',
+  // Add more paths as needed
+};
+
+export async function GET(
```
Comment (Hermione2408) [inline app/api/createFileStructure/route.ts:19]: what is this function about and why it is named as GET only?

File: components/atg/Editor/editor/code.tsx [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+  const [showText, setShowText] = useState<boolean>(false);
+  const testFileRegex = /^test-\d+\.yaml$/; // Regular expression to match 'test-<number>.yaml'
+  const [codeSubmissionId, setSubmissionId] = useState("");
+  const { handleSubmit } = useEditTestSubscription(codeSubmissionId);
+  const code = selectedFile.content;
+  let language = selectedFile.name.split(".").pop();
+  
+  if (language === "js" || language === "jsx") language = "javascript";
+  else if (language === "ts" || language === "tsx") language = "typescript";
+  else if (language === "go") language = "go";
```
Comment (Hermione2408) [inline components/atg/Editor/editor/code.tsx:35]: We can simplify the code by removing the else if condition since we're already assigning "go" as the language in that block.

File: components/atg/Editor/editor/code.tsx [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+  
+        console.log("changed file.");
+  
+        try {
+  
+          console.log("Subscription result", { data, loading, error });
+  
+          if (!loading && data) {
+            // Handle successful subscription data here
+            console.log("Subscription successful:", data);
```
Comment (Hermione2408) [inline components/atg/Editor/editor/code.tsx:63]: why is this log needed?  if case is to notify user, Can we add a Snackbar or some form of notification here to inform the user that the subscription was successful?

**Review by Hermione2408 — COMMENTED** [review]
@Aditya-eddy 
It seems that some of the previous points I mentioned regarding the file-tree component haven't been reflected in the latest changes. Can you please check?

File: components/atg/Editor/editor/code.tsx (hunk 26 above)
Comment (Aditya-eddy) [reply components/atg/Editor/editor/code.tsx:63]: Yes we can add a snackbar saying your test-<number>.yaml has been saved? Does that work?

File: components/atg/Editor/components/file-tree.tsx
Comment (Aditya-eddy) [reply components/atg/Editor/components/file-tree.tsx:1]: I have seprated the components.

File: components/atg/Editor/components/file-tree.tsx (hunk 18 above)
Comment (Aditya-eddy) [reply components/atg/Editor/components/file-tree.tsx]: Done.

**Comment by Sonichigo** [comment]
Me and @khareyash05 have test and looks good to us!

**Comment by khareyash05** [comment]
@Aditya-eddy , we have tested the same, could you also pls test it on your side with the latest landing page changes

**Comment by Aditya-eddy** [comment]
Hey @khareyash05 and @Sonichigo i tested it. And LGTM!!

**Comment by khareyash05** [comment]
Fixed some backend issues and we are GTG!! 🚀
